File: 41. First Missing Positive/41. First Missing Positive.js
function firstMissingPositive(nums) {
    let n = nums.length;

    // Step 1: Clean up the array
    for (let i = 0; i < n; i++) {
        while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] !== nums[i]) {
            // Swap nums[i] with nums[nums[i] - 1]
            let temp = nums[i];
            nums[i] = nums[nums[i] - 1];
            nums[temp - 1] = temp;
        }
    }

    // Step 2: Find the first index with the incorrect number
    for (let i = 0; i < n; i++) {
        if (nums[i] !== i + 1) {
            return i + 1;
        }
    }

    // Step 3: All numbers from 1 to n are present, return n + 1
    return n + 1;
}

File: 41. First Missing Positive/41. First Missing Positive.rb
# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
    n = nums.length
    (0...n).each do |i|
        while nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
        end
    end
    
    (0...n).each do |i|
        return i + 1 if nums[i] != i + 1
    end
    
    n + 1
end